go/libraries/doltcore/sqle/dsess/transactions.go

```diff
@@ -112,62 +117,117 @@ func (tx *DoltTransaction) Commit(ctx *sql.Context, workingSet *doltdb.WorkingSe
 	}
 
 	for i := 0; i < maxTxCommitRetries; i++ {
-		newWorkingSet := false
-
-		ws, err := tx.dbData.Ddb.ResolveWorkingSet(ctx, tx.workingSetRef)
-		if err == doltdb.ErrWorkingSetNotFound {
-			// This is to handle the case where an existing DB pre working sets is committing to this HEAD for the
-			// first time. Can be removed and called an error post 1.0
-			ws = doltdb.EmptyWorkingSet(tx.workingSetRef)
-			newWorkingSet = true
-		} else if err != nil {
-			return nil, err
-		}
+		updatedWs, err := func() (*doltdb.WorkingSet, error) {
+			// Serialize commits, since only one can possibly succeed at a time anyway
+			txLock.Lock()
+			defer txLock.Unlock()
+
+			newWorkingSet := false
+
+			ws, err := tx.dbData.Ddb.ResolveWorkingSet(ctx, tx.workingSetRef)
+			if err == doltdb.ErrWorkingSetNotFound {
+				// This is to handle the case where an existing DB pre working sets is committing to this HEAD for the
+				// first time. Can be removed and called an error post 1.0
+				ws = doltdb.EmptyWorkingSet(tx.workingSetRef)
+				newWorkingSet = true
+			} else if err != nil {
+				return nil, err
+			}
 
-		existingWorkingRoot := ws.WorkingRoot()
+			existingWorkingRoot := ws.WorkingRoot()
 
-		hash, err := ws.HashOf()
-		if err != nil {
-			return nil, err
-		}
+			hash, err := ws.HashOf()
+			if err != nil {
+				return nil, err
+			}
+
+			if newWorkingSet || rootsEqual(existingWorkingRoot, tx.startState.WorkingRoot()) {
+				// ff merge
+				err = tx.dbData.Ddb.UpdateWorkingSet(ctx, tx.workingSetRef, workingSet, hash, tx.getWorkingSetMeta(ctx))
+				if err == datas.ErrOptimisticLockFailed {
+					return nil, nil
```

Comment:
why do we throw away this error? we should avoid this error entirely with serialized commits

Reply:
This is essentially a continue, I'll add a comment